processing/src/main/java/io/druid/query/groupby/having/DimFilterHavingSpec.java

```diff
@@ -54,19 +57,24 @@ public DimFilter getDimFilter()
   }
 
   @Override
-  public boolean eval(final Row row)
+  public void setRowType(Map<String, ValueType> rowType)
   {
-    // Not thread safe, but it doesn't have to be.
-    valueMatcherFactory.setRow(row);
-    try {
-      return valueMatcher.matches();
-    }
-    finally {
-      valueMatcherFactory.setRow(null);
+    synchronized (valueMatcherMonitor) {
```

Comment:
The reason I had for making DimFilterHavingSpec thread-safe is that it's a single object shared by everything processing the groupBy query, and it seemed like making it non-thread-safe was asking for trouble. Some kinds of potential trouble are,

- IntervalChunkingQueryRunner splits up result merging into multiple threads if `chunkPeriod` is set. Currently it _doesn't_ split up the havingSpec processing, since the work is done on accumulate/toYielder rather than on construction of the Sequence, but it seems a flimsy guarantee that things will remain like this forever.
- We may want to implement multithreaded merging of groupBy results in the future, independent of IntervalChunkingQueryRunner, and this may involve HavingSpecs being shared between threads.

But I guess this isn't really necessary, since I do believe that _right now_ it's not shared between threads. So I can make it not thread safe and add a comment about that to HavingSpec.